### PR TITLE
Refactor AuditResults.

### LINKED ISSUE #1965: refactor AuditResults class
From  [maint/tech debt storytime discussion Wed 10/19](https://docs.google.com/document/d/1vTv9ySljSgJ-zcrH2ZztDpdUtnVCxe_hgfZnIehkPGs/edit#heading=h.a6kqlslw7c5o):

This is one of our 5 biggest classes:

image

John already has some ideas for refactors / rethinking of how this is done, but feel free to branch out on your own ...

### PR DESCRIPTION
**Pull request by justinlittman**
closes #1965

## Why was this change made? 🤔
Clearer code? Single responsibility?

## How was this change tested? 🤨

⚡ ⚠ If this change has cross service impact, or if it changes code used internally for cloud replication, ***run [integration test create_preassembly_image_spec.rb](https://github.com/sul-dlss/infrastructure-integration-test/blob/main/spec/features/create_preassembly_image_spec.rb) on stage as it tests preservation***, and/or test in stage environment, in addition to specs. The main classes relevant to replication are `ZipmakerJob`, `PlexerJob`, `*DeliveryJob`, `ResultsRecorderJob`, and `DruidVersionZip`; [see here for overview diagram of replication pipeline](https://github.com/sul-dlss/preservation_catalog/blob/main/app/jobs/README.md).⚡

### CODE REVIEW

File: app/models/audit_results.rb
Comment (mjgiarlo) [inline app/models/audit_results.rb:6]: Is `app/models/` the best place for this new version of the class? I don't know. But I think it's a better fit than `app/services/` and I'm not sure inventing a new `app/` subdirectory is better than stuffing this PORO into `app/models/` so :shipit: ?

File: app/models/audit_results.rb
Comment (mjgiarlo) [inline app/models/audit_results.rb:86]: Appreciate moving to keyword args over positional ones!

File: app/models/audit_results.rb
Comment (mjgiarlo) [inline app/models/audit_results.rb:139]: Did this method move somewhere or is it just nuked? Asking because I think it's still invoked elsewhere in the `AuditResults` model.

File: app/models/audit_results.rb
Comment (mjgiarlo) [inline app/models/audit_results.rb:129]: Appreciate being intentional about what's in the public API and what's an implementation detail.

**Review by mjgiarlo — COMMENTED** [review]
Looks better to me.

File: app/models/audit_results.rb
Comment (ndushay) [reply app/models/audit_results.rb:6]: I think this is how stuff ended up in `app/lib` -- because it wasn't a rails model, which is all I expect to find in the `app/models` directory.   I'm not sure what's best;  Putting it in the services junk drawer feels more comfortable to me.  :shrug:   I would defer to John.

File: app/models/audit_results.rb
Comment (ndushay) [inline app/models/audit_results.rb:14]: I'm not in love with these comments, but they are informative.  What is the thinking behind removing them ... and not adding this info anywhere else (e.g. in the "beginners guide" that doesn't exists yet ... but there is a ticket that could get this info pasted as a comment)

File: app/models/audit_results.rb
Comment (ndushay) [reply app/models/audit_results.rb:86]: Maybe I'm being slow, but ... who are the callers and are these changes made there?

File: app/services/audit_results_reporter.rb
Comment (ndushay) [inline app/services/audit_results_reporter.rb:4]: Here, I like breaking this out ... but isn't it also a service that is only used by one class?  That is the argument used for NOT breaking a service into a separate class in a different PR.   

Can you say why this is a different case, or something?  I am a bit baffled.

File: app/models/audit_results.rb
Comment (mjgiarlo) [reply app/models/audit_results.rb:6]: @ndushay That'd work for me too. 😄

File: app/models/audit_results.rb
Comment (mjgiarlo) [reply app/models/audit_results.rb:86]: You're not being slow; this isn't done yet (AFAICT). I don't think this PR has a passing test suite yet. My vague recollection is J-Litt wanted some early feedback on decisions made so far before proceeding further.

File: app/models/audit_results.rb
Comment (mjgiarlo) [reply app/models/audit_results.rb:14]: Good call.

File: app/models/audit_results.rb
Comment (justinlittman) [reply app/models/audit_results.rb:6]: In our other apps, we have non-Rails models in `app/models/`. At least based on my familiarity with our other apps, that is where I would have looked for it.

File: app/models/audit_results.rb
Comment (justinlittman) [reply app/models/audit_results.rb:86]: Correct. I didn't want to change all of the callers and tests until there was consensus on the refactor/split.

File: app/models/audit_results.rb
Comment (justinlittman) [reply app/models/audit_results.rb:112]: Yes, the previous implementation allowed other code to manipulate the result array. The errs on the other side, and makes sure that other code knows it cannot manipulate the results.

File: app/models/audit_results.rb
Comment (justinlittman) [reply app/models/audit_results.rb:139]: Ooops. You're right.

File: app/services/audit_results_reporter.rb
Comment (justinlittman) [reply app/services/audit_results_reporter.rb:4]: 1. This class isn't used by only one class. (It isn't invoked in `AuditResult` at all.)
2. It will have completely separate tests from `AuditResult`.

File: app/models/audit_results.rb
Comment (justinlittman) [reply app/models/audit_results.rb:14]: I put some of this back in.

File: app/models/audit_results.rb
Comment (mjgiarlo) [reply app/models/audit_results.rb:112]: Excellent.

File: app/models/audit_results.rb
Comment (mjgiarlo) [reply app/models/audit_results.rb:6]: I think you've convinced me `app/models/` is fine here.

File: app/models/audit_results.rb
Comment (mjgiarlo) [inline app/models/audit_results.rb:137]: If we're merely checking for the presence of a match, without caring about the value of the match, I think this method could be:

```suggestion
  def status_changed_to_ok?(result)
    /to ok$/.match?(result[AuditResults::CM_STATUS_CHANGED])
```

**Review by mjgiarlo — APPROVED** [review]
Approve pending consideration of one minor tweak.

File: app/models/audit_results.rb
Comment (justinlittman) [reply app/models/audit_results.rb:137]: True, but I would rather the method return a boolean instead of a Matcher.

Comment (mjgiarlo) [reply app/models/audit_results.rb:137]: @justinlittman `#match?` returns a boolean. (`#match` doesn't.)

Comment (justinlittman) [reply app/models/audit_results.rb:137]: Super. I'll make that change.

**Pull request merged**
